File: subly_ios/Subly/Subly/Domain/Repositories/SubscriptionRepository.swift
import Foundation
import Combine

protocol SubscriptionRepository {
    func getAllSubscriptions() -> AnyPublisher<[Subscription], Never>
    func getActiveSubscriptions() -> AnyPublisher<[Subscription], Never>
    func getSubscriptionById(_ id: UUID) -> AnyPublisher<Subscription?, Never>
    func getSubscriptionsBetweenDates(start: Date, end: Date) -> AnyPublisher<[Subscription], Never>
    func getUpcomingSubscriptions(before date: Date) -> AnyPublisher<[Subscription], Never>
    func getMonthlyTotal() -> AnyPublisher<Double, Never>
    func insertSubscription(_ subscription: Subscription) async throws
    func updateSubscription(_ subscription: Subscription) async throws
    func deleteSubscription(_ subscription: Subscription) async throws
    func deleteSubscription(id: UUID) async throws
}

File: subly_android/app/src/main/java/net/cynreub/subly/domain/repository/SubscriptionRepository.kt
package net.cynreub.subly.domain.repository

import kotlinx.coroutines.flow.Flow
import net.cynreub.subly.domain.model.Subscription
import java.time.LocalDate
import java.util.UUID

interface SubscriptionRepository {
    fun getAllSubscriptions(): Flow<List<Subscription>>
    fun getActiveSubscriptions(): Flow<List<Subscription>>
    fun getSubscriptionById(id: UUID): Flow<Subscription?>
    fun getSubscriptionsBetweenDates(startDate: LocalDate, endDate: LocalDate): Flow<List<Subscription>>
    fun getUpcomingSubscriptions(date: LocalDate): Flow<List<Subscription>>
    fun getMonthlyTotal(): Flow<Double>
    suspend fun insertSubscription(subscription: Subscription)
    suspend fun updateSubscription(subscription: Subscription)
    suspend fun deleteSubscription(subscription: Subscription)
    suspend fun deleteSubscriptionById(id: UUID)
}
